docs/src/content/docs/book/functions.mdx

```diff
@@ -114,6 +116,243 @@ native storeUint(s: Builder, value: Int, bits: Int): Builder;
+//                |       sits on top of the stack
+//                Result pushed first by LDIX,
+//                sits on the bottom of the stack
+```
+
+Thus, when the results are popped from the stack, the [Struct][struct] fields are filled from right to left: the top stack value is mapped to the last field, the second-to-top value is mapped to the second-to-last field, and so on.
+
+Note, that when using [Structs][struct] as the return type, you must map all stack values onto respective Struct fields, otherwise an error with [exit code 7](/book/exit-codes#7) will be thrown: `Type check error`.
+
+If you only need the bottom value from the stack, it is possible to specify a [primitive type][p] as the return value instead:
+
+```tact
+// Same function as before, but now we don't use the `IntSlice` Struct
+// and instead only take the bottom result sitting in the stack
+asm fun sliceLoadInt(s: Slice, len: Int): Int { LDIX }
+//                                        ↑
+//                                        captures the Int value, discarding
+//                                        the Slice one produced by LDIX instruction
+```
+
+### Stack registers {#asm-stack-registers}
+
+The so-called _stack registers_ are conventional way of referring to the values at the top of the stack. In total, there are $256$ stack registers, i.e. values held on the stack at any given time. You can refer to any of them using any of `s0`, `s1`, ..., `s255`.
+
+Register `s0` is the value at the top of the stack, register `s1` is the value immediately after it, and so on, until we reach the bottom of the stack, represented by `s255`, i.e. the $256$th stack register. When a value `x` is pushed onto a stack, it becomes the new `s0`. At the same time, old `s0` becomes new `s1`, old `s1` — new `s2`, and so on.
+
+```tact
+asm fun takeSecond(a: Int, b: Int): Int {
+    //             ↑       ↑
+    //             |       Pushed last, sits on top of the stack
+    //             Pushed first, sits second from the top of the stack
+
+    // Now, let's swap the s0 (top of the stack) with s1 (second-to-top):
+    s1 XCHG0
+
+    // Then, let's drop the value from the top of the stack
+    DROP
+
+    // At the end, we have only one value on the stack, which is b
+}
+
+fun showcase() {
+    takeFirst(5, 10); // 10, i.e. b
+}
+```
+
+### Arrangements {#asm-arrangements}
+
+Often times it's useful to change the order of arguments pushed to the stack or the order of return values without referring to stack registers in the body. You can do that with `asm{:tact}` arrangements in the following manner:
+
+```tact
+// Changing the order of arguments to match the STDICT signature:
+// `c` will be pushed first and get on the bottom of the stack,
+// while `self` will be pushed last and get on top of the stack
+asm(c self) extends fun asmStoreDict(self: Builder, c: Cell?): Builder { STDICT }
+
+// Changing the order of return values of LDVARUINT16,
```

Comment:
It is still not clear what the notation `-> 1 0` means regarding what happens to the results of `LDVARUINT16` **in** the stack itself. The explanation states that 1 represents the value of stack register 1, etc. but it does not explain the significance of writing them in the order `-> 1 0`. Probably what needs to be said is that the notation `-> 1 0` describes how the contents of the stack will be rearranged, when reading `-> 1 0` left-to-right: the contents of register `s1` will be placed at the top of the stack, and the contents of register `s0` will be placed second-to-top. 

One alternative way of explaining could be in terms of removing from the stack: `-> 1 0` means that `s1` is removed first, followed by `s0`.  Hence, the function returns the `Builder` in `s0` because it was the stack content removed last.

Reply:
Now I am having second thoughts on using "removing" because it becomes confusing with what happens with the rest of the stack. For example, suppose that after executing some asm function with declaration `-> 2 1 0`, we have the 5 element stack (top is leftmost):
```
a b c d e
```
Then, `-> 0 1 2` means "remove `s0`, then `s1`, then `s2`", so that the stack after removing `s0` is:
```
b c d e
```
But then, `s0` contains now `b`, when previously `b` was in `s1`.

So, probably a better word instead of "removing" would be "read from":

`-> 1 0` means that `s1` is read from the stack first, followed by `s0`. Hence, the function returns the `Builder` in `s0` because it was the stack content read last.

Reply:
Thing is, as I just checked in tests, `-> 0 1 2` is not about taking or not taking any results, but merely about positioning items for the whatever result type we've specified. Like, if the return type is `Int`, one can only specify `-> 0` and nothing else, even though `-> 0` in this case is the same as not writing anything at all. And when the Structs, long Structs (more than 15 entries) or even nested Structs are involved, this is getting complicated.

Thus, my description of `s0` matching `0`, `s1` matching `1` is actually incorrect and has to be rewritten. And I've got to check the cases with long or nested Structs here as well, same as for the "stack calling conventions" bit.

Reply:
I see. So, this declaration is incorrect (because it returns only one element):
```
asm(self len -> 1 0) extends fun asmLoadInt(self: Slice, len: Int): Slice { LDIX }
```
but this is correct:
```
asm(self len) extends fun asmLoadInt(self: Slice, len: Int): Int { LDIX }
```
even though it will discard the `Slice` result and keep only the `Int`. Or is this last one also incorrect?

Mmmm.... very confusing indeed. So, when using the notation `-> m n p` it is not possible to discard values in the result type. I think this is acceptable. It is better to explicitly state all the results than to rely on understanding implicit discards.

Reply:
First one is incorrect. Second one could've been correct if we had our own backend or if we'd alter FunC generation, but since I tested that it's also incorrect — nothing can be discarded in result type.

It worked for me in previous tests mainly because FunC doesn't any checks and when all asm function bodies are embedded in Fift code, and I had some DROP instructions very deep later on in other asm functions, which unexpectedly (for me) cleared the stack for this one. And I noticed that a little too late.

In the end, this really proves the point of those cautionary paragraphs at the top of the assembly functions description. This stuff is really messy, intertwined and hard to debug. But I'll persevere.

Reply:
> First one is incorrect. Second one could've been correct if we had our own backend or if we'd alter FunC generation, but since I tested that it's also incorrect — nothing can be discarded in result type.

Now that I've tested more things I can say that discarding is possible, but works somewhat stable only for integers. Discarding cells causes all kinds of exceptions all over the place.

Here's an example:

```tact
struct TestNest { a: Int; b: TestBig; c: Int }
struct TestBig { f1: Int; f2: Int; f3: Int; f4: Int; f5: Int }

asm fun test(structure: TestNest): Int {
    DROP DROP
}

fun oneThing() {
    dump(
        test(
            TestNest{
                a: 1,
                b: TestBig{ f1: 2, f2: 3, f3: 4, f4: 5, f5: 6 },
                c: 100,
            }
        )
    ); // → 5, since the 100 and 6 were removed from the top of the stack by two DROPs
}
```

Reply:
Btw, going beyond 16 fields in Structs fails the compilation of FunC. @Gusarich, wasn't that issue resolved long time ago? Or, perhaps, the way we transpile asm functions differs from the rest of the functions that you've fixed?

Reply:
Very dangerous indeed. In the future, once we have our embedded assembly language we will be able to add checks to asm functions, as Anton mentioned in a previous comment. For the moment, we just need to add to the documentation the extra stuff you mentioned (with examples would be better):
- Flattening of structs in arguments and results.
- Limit to 16 arguments and 16 results (counting flattened structs).
- Effect of mutating functions on the stack.

I think the limit of 16 arguments and results is not that bad. In principle, asm functions are used to program very specific, small procedures with simple logic. Is there anyone that would program almost the entire contract logic using asm instructions? I highly doubt it. I mean, the point of using Tact is precisely not to write everything using assembly language. But I agree that in the future, the limit of 16 could be changed.

Reply:
Packing notes from the weekend to finish this PR today ASAP.